apps/studio/src/server/modules/user/inactiveUsers.service.ts

```diff
@@ -0,0 +1,198 @@
+  toDaysAgo?: number
+}
+export const getInactiveUsers = async ({
+  fromDaysAgo,
+  toDaysAgo = MAX_DAYS_FROM_LAST_LOGIN,
+}: GetInactiveUsersProps): Promise<User[]> => {
+  const fromDateThreshold = fromDaysAgo ? getDateOnlyInSG(fromDaysAgo) : null
+  const toDateThreshold = getDateOnlyInSG(toDaysAgo)
+
+  return db
+    .selectFrom("User")
+    .innerJoin("ResourcePermission", "ResourcePermission.userId", "User.id")
+    .where("User.deletedAt", "is", null)
+    .where("ResourcePermission.deletedAt", "is", null)
+    .where("User.email", "not in", ISOMER_ADMINS_AND_MIGRATORS_EMAILS) // needed to provide support for agencies
+    .where((eb) =>
+      eb.or([
+        // Users who have never logged in
+        eb.and([
+          eb("User.lastLoginAt", "is", null),
+          eb("User.createdAt", "<=", toDateThreshold),
+          ...(fromDateThreshold
+            ? [eb("User.createdAt", ">", fromDateThreshold)]
+            : []),
+        ]),
+        // Users who have logged in but haven't logged in for a while
+        eb.and([
+          eb("User.lastLoginAt", "is not", null),
+          eb("User.lastLoginAt", "<=", toDateThreshold),
+          ...(fromDateThreshold
+            ? [eb("User.lastLoginAt", ">", fromDateThreshold)]
+            : []),
+        ]),
+      ]),
+    )
+    .selectAll(["User"])
+    .distinct()
+    .execute()
+}
+
+interface DeactivateUserProps {
+  user: User
+  userIdsToDeactivate: User["id"][] // used for idempotency purposes
+}
+export const deactivateUser = async ({
+  user,
+  userIdsToDeactivate,
+}: DeactivateUserProps): Promise<void> => {
+  // This should not happen as this sub-function is only called when there are users to deactivate
+  // Nevertheless, needed to add this as a safety net to ensure valid Kysely query
+  if (userIdsToDeactivate.length === 0) return
+
+  // Note: we are just deleting their site permissions (by setting deletedAt)
+  // we are NOT deleting the user themselves because the intention is to
+  // remove their access to interact with their sites, not remove them from Isomer
+  let sitesAndAdmins: AccountDeactivationEmailTemplateData["sitesAndAdmins"] =
+    []
+
+  try {
+    sitesAndAdmins = await db
+      .transaction()
+      .setIsolationLevel("serializable")
+      .execute(async (tx) => {
+        const deletedPermissions = await tx
+          .updateTable("ResourcePermission")
+          .where("userId", "=", user.id)
+          .where("deletedAt", "is", null)
+          .set({ deletedAt: new Date() })
+          .returningAll()
+          .execute()
+
+        // For idempotency purposes since multiple nodes may run this function at the same time
+        if (deletedPermissions.length === 0) return []
+
+        const siteIdsUserHasPermissionsFor = deletedPermissions.map(
+          (p) => p.siteId,
+        )
+
+        return await tx
+          .with("siteAdmins", (eb) =>
+            eb
+              .selectFrom("Site")
+              .leftJoin(
```

Comment:
[question] why is this a `leftJoin`? the left table is `Site` and this just means that there will be sites without a joined `ResourcePermission`?